payu/git_utils.py

```diff
@@ -0,0 +1,173 @@
+"""Simple wrappers around git commands
+
+Using the GitPython library for interacting with Git
+"""
+
+import warnings
+from pathlib import Path
+from typing import Optional, Union, List, Dict
+
+import git
+import configparser
+
+
+class PayuBranchError(Exception):
+    """Custom exception for payu branch operations"""
+
+
+class PayuGitWarning(Warning):
+    """Custom warning class - useful for testing"""
+
+
+def get_git_repository(repo_path: Union[Path, str],
+                       initialise: bool = False,
+                       catch_error: bool = False) -> Optional[git.Repo]:
+    """Return a PythonGit repository object at given path. If initialise is
+    true, it will attempt to initialise a repository if it does not exist.
+    Otherwise, if catch_error is true, it will return None"""
+    try:
+        repo = git.Repo(repo_path)
+        return repo
+    except git.exc.InvalidGitRepositoryError:
+        if initialise:
+            repo = git.Repo.init(repo_path)
+            print(f"Initialised new git repository at: {repo_path}")
+            return repo
+
+        warnings.warn(
+            f"Path is not a valid git repository: {repo_path}",
+            PayuGitWarning
+        )
+        if catch_error:
+            return None
+        raise
+
+
+def get_git_branch(repo_path: Union[Path, str]) -> Optional[str]:
+    """Return the current git branch or None if repository path is not a git
+    repository"""
+    repo = get_git_repository(repo_path, catch_error=True)
+    if repo:
+        return str(repo.active_branch)
+
+
+def get_git_user_info(repo_path: Union[Path, str],
+                      config_key: str,
+                      example_value: str) -> Optional[str]:
+    """Return git config user info, None otherwise. Used for retrieving
+    name and email saved in git"""
+    repo = get_git_repository(repo_path, catch_error=True)
+    if repo is None:
+        return
+
+    try:
+        user_value = repo.config_reader().get_value('user', config_key)
+        return user_value
+    except (configparser.NoSectionError, configparser.NoOptionError):
+        print(
+            f'No git config set for user.{config_key}. '
+            'To set run the following inside the control repository:\n'
+            f'    git config user.{config_key} "{example_value}"'
+        )
+
+
+def git_commit(repo_path: Union[Path, str],
+               commit_message: str,
+               paths_to_commit: List[Union[Path, str]]) -> None:
+    """Add a git commit of changes to paths"""
+    # Get/Create git repository - initialise is true as adding a commit
+    # directly after
+    repo = get_git_repository(repo_path, initialise=True)
+
+    # Un-stage any pre-existing changes
+    repo.index.reset()
```

Comment:
> Yeah I thinking to avoid adding unrelated changes so the metadata commit was only adding the metadata file. If it was being run as part of setup/run, those changes would instead get added and committed on the runlog commit.

Is that an argument then to `stash` changes, make the changes to the metadata, commit and then `stash pop` the changed state back again? Having said which, `git stash` [can be a little tricky](https://stackoverflow.com/a/31596366).

> Would preventing checkout branch with staged/un staged changes be a good idea or add in a warning as it could be useful to move new changes to new branch?

I think a warning would be a good idea. I tend to think preventing the checkout with staged changes sounds like a good idea. Unstaged changes shouldn't be an issue, as the commit is only add `metadata.yaml` right? Or do you mean unstaged changes to `metadata.yaml`?

> We also spoke about always committing metadata file if existing repository was a git repository - even if `runlog` is false (which I'll need to add in), as there were only be one commit per branch. If it was common to merge development branches into one branch then that could be an issue - the merged branch will have a broken experiment name which includes the branch that was added metadata.. This will be picked up on the next `payu setup` and could be fixed by removing the `experiment` field from `metadata.yaml`

Ugh. Yeah it could be a bit messy eh?

Ok, what happens if we don't commit `metadata.yaml`?

Reply:
> Ugh. Yeah it could be a bit messy eh?
>Ok, what happens if we don't commit metadata.yaml?

Chaos.. If metadata isn't committed, when you switch to a different branch, that metadata file will carry over with a different branch name in the `experiment` name field - which will then raise errors in the next `payu setup/run`. 

I am wondering whether it'll be worth adding a config flag, which flags to not create/commit metadata file (or use `runlog: False` or `metadata: enable: False`). An option in this case could be to use `${CONTROL_DIR}-${BRANCH}` instead for the `experiment` name, unless there's an existing legacy archive, then `${CONTROL_DIR}` for the experiment name and to determine this at `payu run/setup`. Or just to use the old workflow so `${CONTROL_DIR}` for everything if the flag is true/false and not commit/create metadata file. 

Another option is having the existing `experiment` in config as an override for the experiment field in metadata - in this case the responsibility for different experiment names for branches is on the developer.

Alternatively larger change option to is remove `experiment` from metadata completely and just dynamically determine experiment name at runtime. The downside of this is you can't mix legacy experiments with new branch-id-aware experiments in the same control repository. Though now thinking about it `payu checkout -b` could create an empty archive folder with a new branch-uuid aware `experiment` name... 
Then in `payu setup/run/sweep`'s Experiment initialisation: 
1. if branch-id-aware archive exists, `experiment.name = ${CONTROL_DIR}-${BRANCH}-${UUID}`
2. if legacy `experiment` archive exists, `experiment.name = ${CONTROL_DIR}`
3. else no archive - generate new uuid and create an empty archive and set  `experiment.name = ${CONTROL_DIR}-${BRANCH}-${UUID}` 
`payu checkout` - then becomes simpler.  If `payu checkout -b` - generates new uuid and create an empty archive with branch-uuid aware name, remove/add symlinks. Otherwise `payu checkout` just switches symlinks. 

Merging branches with different `uuid`s in metadata would could still change `uuid` but at that point the next `payu setup`, if `${CONTROL_DIR}-${BRANCH}-${UUID}` or `${CONTROL_DIR}` archive doesn't exist, it'll generate a new uuid and archive (and let the user know that was it has done..). If the uuid change wasn't desired,  could directly modify `uuid` in `config.yaml` to revert it to a previous one. 

Sorry the idea of creating an empty archive folder for new branch-id aware experiment did not occur to me earlier.. In this way, similarly to the code in this PR- as long as you create branches using `payu checkout -b`, you can add branch-aware experiments to the same control directories that have a legacy archive experiment..

Reply:
So you're suggesting to not have the experiment in `metadata.yaml` but dynamically determine it when it is required.

This has the advantage of dramatically reducing the number of commits required for `metadata.yaml`? Only when `uuid` changes?

Sounds good to me! Is this a lot of work to do? 

What does this mean for committing `metadata.yaml`? Do we do it regardless of `runlog` setting now that there are fewer commits? I think it sounds like a good idea.

Reply:
It'll reduced the commits if `--keep-uuid` is used for development branches/branches that will be merged. This will give only only 1 commit to `metadata.yaml` - as branches that will be merged will share one uuid. e.g.  say `metadata.yaml` was:
````
uuid: 1234567890
````


````
Branch: BRANCH1
	# experiment: CONTROL-BRANCH1-12345
	uuid: 1234567890
Branch: BRANCH2
	# experiment: CONTROL-BRANCH2-12345
	uuid: 1234567890
````

`payu checkout` to move between branches and set up symlinks. 
`payu checkout -b NEW_BRANCH --keep-uuid` Leaves keeps uuid when moving to new branch.

Using `keep-uuid` is not recommended workflow for experiment tracking as uuid is unchanged.

Reply:
> Sounds good to me! Is this a lot of work to do?

I have most of the changes already done on my local branch. It's working apart from now the logic has changed a bit, warnings are being raised in other tests. I'll then need to fix up the documentation to reflect those changes. So it's doable...